Index: universe/Names.h
===================================================================
--- universe/Names.h	(revision 0)
+++ universe/Names.h	(revision 0)
@@ -0,0 +1,247 @@
+// -*- C++ -*-
+#ifndef _Names_h_
+#define _Names_h_
+
+#include <GG/adobe/name.hpp>
+
+#include <boost/preprocessor/cat.hpp>
+#include <boost/preprocessor/seq/for_each.hpp>
+
+
+#define NAMES_SEQ                               \
+    (Activation)                                \
+    (AddedBefore)                               \
+    (AddedSince)                                \
+    (AddSpecial)                                \
+    (AddStarlanes)                              \
+    (Affiliation)                               \
+    (Age)                                       \
+    (Alignment)                                 \
+    (AlignmentEffects)                          \
+    (All)                                       \
+    (And)                                       \
+    (AntiFighterDamage)                         \
+    (AntiShipDamage)                            \
+    (Armed)                                     \
+    (Asteroids)                                 \
+    (BattleSpeed)                               \
+    (BuildCost)                                 \
+    (Building)                                  \
+    (BuildingType)                              \
+    (BuildTime)                                 \
+    (CanColonize)                               \
+    (CanProduceShips)                           \
+    (Capacity)                                  \
+    (Capital)                                   \
+    (CaptureResult)                             \
+    (Category)                                  \
+    (Class)                                     \
+    (Colour)                                    \
+    (Condition)                                 \
+    (Construction)                              \
+    (ContainedBy)                               \
+    (Contains)                                  \
+    (CreateBuilding)                            \
+    (CreatedOnTurn)                             \
+    (CreatePlanet)                              \
+    (CreateShip)                                \
+    (CreationTurn)                              \
+    (CurrentTurn)                               \
+    (Damage)                                    \
+    (Data)                                      \
+    (Defense)                                   \
+    (Description)                               \
+    (Design)                                    \
+    (DesignHasHull)                             \
+    (DesignHasPart)                             \
+    (DesignHasPartClass)                        \
+    (DesignID)                                  \
+    (DesignName)                                \
+    (Destination)                               \
+    (Destroy)                                   \
+    (Detection)                                 \
+    (Distance)                                  \
+    (DistanceToSource)                          \
+    (Effects)                                   \
+    (EffectsGroup)                              \
+    (EffectsGroups)                             \
+    (Empire)                                    \
+    (Endpoint)                                  \
+    (Environment)                               \
+    (Environments)                              \
+    (ExploredByEmpire)                          \
+    (Farming)                                   \
+    (FighterType)                               \
+    (FighterWeaponRange)                        \
+    (FinalDestinationID)                        \
+    (Fleet)                                     \
+    (FleetID)                                   \
+    (FleetSupplyableByEmpire)                   \
+    (Foci)                                      \
+    (Focus)                                     \
+    (FocusType)                                 \
+    (FoodConsumption)                           \
+    (FoodStockpile)                             \
+    (Fuel)                                      \
+    (GasGiant)                                  \
+    (GenerateSitrepMessage)                     \
+    (GiveTechToOwner)                           \
+    (Graphic)                                   \
+    (HasSpecial)                                \
+    (HasSpecialSinceTurn)                       \
+    (Health)                                    \
+    (High)                                      \
+    (Homeworld)                                 \
+    (Hull)                                      \
+    (ID)                                        \
+    (Industry)                                  \
+    (InSystem)                                  \
+    (Item)                                      \
+    (Jumps)                                     \
+    (Label)                                     \
+    (LaunchRate)                                \
+    (LocalCandidate)                            \
+    (Location)                                  \
+    (LockTechItemsForOwner)                     \
+    (Lookup_Strings)                            \
+    (Low)                                       \
+    (MaxDefense)                                \
+    (MaxFuel)                                   \
+    (MaximumNumberOf)                           \
+    (MaxShield)                                 \
+    (MaxStructure)                              \
+    (MaxTroops)                                 \
+    (Message)                                   \
+    (Meter)                                     \
+    (MineralStockpile)                          \
+    (MinimumNumberOf)                           \
+    (Mining)                                    \
+    (Mode)                                      \
+    (Model)                                     \
+    (ModeNumberOf)                              \
+    (Monster)                                   \
+    (MonsterFleet)                              \
+    (MountableSlotTypes)                        \
+    (MoveTo)                                    \
+    (Name)                                      \
+    (NextBetterPlanetType)                      \
+    (NextSystemID)                              \
+    (Not)                                       \
+    (Number)                                    \
+    (NumberOf)                                  \
+    (NumShips)                                  \
+    (Object)                                    \
+    (ObjectType)                                \
+    (Or)                                        \
+    (OwnedBy)                                   \
+    (Owner)                                     \
+    (OwnerFoodStockpile)                        \
+    (OwnerHasTech)                              \
+    (OwnerMineralStockpile)                     \
+    (OwnerTradeStockpile)                       \
+    (Parameters)                                \
+    (Part)                                      \
+    (PartClass)                                 \
+    (PartName)                                  \
+    (Parts)                                     \
+    (Planet)                                    \
+    (Planetbound)                               \
+    (PlanetEnvironment)                         \
+    (PlanetID)                                  \
+    (PlanetSize)                                \
+    (PlanetType)                                \
+    (Playable)                                  \
+    (Population)                                \
+    (Position)                                  \
+    (Prerequisites)                             \
+    (PreviousSystemID)                          \
+    (Probability)                               \
+    (ProducedByEmpire)                          \
+    (ProducedByEmpireID)                        \
+    (Producible)                                \
+    (Property)                                  \
+    (Random)                                    \
+    (Range)                                     \
+    (Reason)                                    \
+    (RemoveSpecial)                             \
+    (RemoveStarlanes)                           \
+    (Research)                                  \
+    (Researchable)                              \
+    (ResearchCost)                              \
+    (ResearchTurns)                             \
+    (ResourceSupplyConnected)                   \
+    (RevokeTechFromOwner)                       \
+    (ROF)                                       \
+    (RootCandidate)                             \
+    (Scope)                                     \
+    (Set)                                       \
+    (SetDestination)                            \
+    (SetEmpireCapital)                          \
+    (SetEmpireFoodStockpile)                    \
+    (SetEmpireMeter)                            \
+    (SetEmpireMineralStockpile)                 \
+    (SetEmpireTradeStockpile)                   \
+    (SetOwner)                                  \
+    (SetPlanetSize)                             \
+    (SetPlanetType)                             \
+    (SetSpecies)                                \
+    (SetStarType)                               \
+    (Shield)                                    \
+    (ShipDesign)                                \
+    (Ships)                                     \
+    (Short_Description)                         \
+    (Size)                                      \
+    (Slot)                                      \
+    (Slots)                                     \
+    (SlotType)                                  \
+    (SortBy)                                    \
+    (SortKey)                                   \
+    (Source)                                    \
+    (Spacebound)                                \
+    (SpawnLimit)                                \
+    (SpawnRate)                                 \
+    (Special)                                   \
+    (Species)                                   \
+    (Speed)                                     \
+    (StackingGroup)                             \
+    (Star)                                      \
+    (StarlaneSpeed)                             \
+    (StarType)                                  \
+    (Stationary)                                \
+    (Stealth)                                   \
+    (Structure)                                 \
+    (Supply)                                    \
+    (System)                                    \
+    (SystemID)                                  \
+    (Tag)                                       \
+    (Target)                                    \
+    (TargetConstruction)                        \
+    (TargetFarming)                             \
+    (TargetHealth)                              \
+    (TargetIndustry)                            \
+    (TargetMining)                              \
+    (TargetPopulation)                          \
+    (TargetResearch)                            \
+    (TargetTrade)                               \
+    (Tech)                                      \
+    (TechType)                                  \
+    (Trade)                                     \
+    (TradeStockpile)                            \
+    (Troops)                                    \
+    (Turn)                                      \
+    (Type)                                      \
+    (Unlock)                                    \
+    (UnlockTechItemsForOwner)                   \
+    (Unproducible)                              \
+    (Unresearchable)                            \
+    (Value)                                     \
+    (Victory)                                   \
+    (VisibleToEmpire)                           \
+    (WithinDistance)                            \
+    (WithinStarlaneJumps)
+
+#define DECLARE_NAME(r, _, elem) extern adobe::aggregate_name_t BOOST_PP_CAT(elem, _name);
+BOOST_PP_SEQ_FOR_EACH(DECLARE_NAME, _, NAMES_SEQ)
+#undef DECLARE_NAME
+
+#endif
Index: universe/Names.cpp
===================================================================
--- universe/Names.cpp	(revision 0)
+++ universe/Names.cpp	(revision 0)
@@ -0,0 +1,8 @@
+#include "Names.h"
+
+#include <boost/preprocessor/stringize.hpp>
+
+
+#define DEFINE_NAME(r, _, name) adobe::aggregate_name_t BOOST_PP_CAT(name, _name) = { BOOST_PP_STRINGIZE(name) };
+BOOST_PP_SEQ_FOR_EACH(DEFINE_NAME, _, NAMES_SEQ)
+#undef DEFINE_NAME
Index: universe/ValueRef.h
===================================================================
--- universe/ValueRef.h	(revision 4463)
+++ universe/ValueRef.h	(working copy)
@@ -7,6 +7,8 @@
 #include "Condition.h"
 #include "../util/MultiplayerCommon.h"
 
+#include <GG/adobe/name.hpp>
+
 #include <boost/algorithm/string/case_conv.hpp>
 #include <boost/format.hpp>
 #include <boost/mpl/if.hpp>
@@ -130,6 +132,8 @@
       * otherwise, the same field is read from Eval's \a target parameter. */
     Variable(ReferenceType ref_type, const std::string& property_name);
 
+    Variable(const std::vector<adobe::name_t>& property_name) {}
+
     ReferenceType                   GetReferenceType() const;
     const std::vector<std::string>& PropertyName() const;
 
@@ -166,6 +170,12 @@
               StatisticType stat_type,
               const Condition::ConditionBase* sampling_condition);
 
+    Statistic(const std::vector<adobe::name_t>& property_name,
+              StatisticType stat_type,
+              const Condition::ConditionBase* sampling_condition) :
+        ValueRef::Variable<T>(property_name)
+        {}
+
     StatisticType                   GetStatisticType() const;
     const Condition::ConditionBase* SamplingCondition() const;
 
Index: universe/ValueRefFwd.h
===================================================================
--- universe/ValueRefFwd.h	(revision 4463)
+++ universe/ValueRefFwd.h	(working copy)
@@ -19,6 +19,7 @@
     template <class T> struct Variable;
     template <class T> struct Statistic;
     enum StatisticType {
+        INVALID_STATISTIC_TYPE = -1,
         COUNT,  // returns the number of objects matching the condition
         SUM,    // returns the sum of the property values of all objects matching the condition
         MEAN,   // returns the mean of the property values of all objects matching the condition
Index: universe/Tech.h
===================================================================
--- universe/Tech.h	(revision 4463)
+++ universe/Tech.h	(working copy)
@@ -141,7 +141,7 @@
     given a set of currently-known techs. */
 class TechManager
 {
-private:
+public:
     struct CategoryIndex {};
     struct NameIndex {};
     typedef boost::multi_index_container<
@@ -166,7 +166,6 @@
         >
     > TechContainer;
 
-public:
     /** iterator that runs over techs within a category */
     typedef TechContainer::index<CategoryIndex>::type::const_iterator category_iterator;
 
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 4463)
+++ CMakeLists.txt	(working copy)
@@ -329,10 +329,17 @@
 ########################################
 # Recurse Into Sources                 #
 ########################################
-add_subdirectory(server)
-add_subdirectory(client/AI)
-add_subdirectory(client/human)
+#add_subdirectory(server)
+#add_subdirectory(client/AI)
+#add_subdirectory(client/human)
 
+option(BUILD_PARSER_TESTS "Controls generation of parser unit tests." OFF)
+
+if (BUILD_PARSER_TESTS)
+    enable_testing()
+    add_subdirectory(parse)
+endif ()
+
 ########################################
 # Win32 SDK-only steps                 #
 ########################################
